### ISSUE TITLE
Cannot download from Apple podcast any more

### ISSUE THREAD

**Issue by sharkshark123** [yt-dlp/yt-dlp#10809]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

taiwan

### Provide a description that is worded well enough to be understood

This is the error message I get

→ ./yt-dlp_macos "https://podcasts.apple.com/tw/podcast/ep480/id1500839292?i=1000665698691"
[ApplePodcasts] Extracting URL: https://podcasts.apple.com/tw/podcast/ep480/id1500839292?i=1000665698691
[ApplePodcasts] 1000665698691: Downloading webpage
ERROR: 1000665698691: An extractor error has occurred. (caused by StopIteration()); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

this is my version

→ ./yt-dlp_macos -U
Latest version: master@2024.08.19.221301 from yt-dlp/yt-dlp-master-builds
yt-dlp is up to date (master@2024.08.19.221301 from yt-dlp/yt-dlp-master-builds)

Was able to download, stop working all of a sudden recently. 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://podcasts.apple.com/tw/podcast/ep480/id1500839292?i=1000665698691']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version master@2024.08.19.221301 from yt-dlp/yt-dlp-master-builds [f0bb28504] (darwin_exe)
[debug] Python 3.12.4 (CPython arm64 64bit) - macOS-14.5-arm64-arm-64bit (OpenSSL 3.0.13 30 Jan 2024)
[debug] exe versions: none
[debug] Optional libraries: Cryptodome-3.20.0, brotli-1.1.0, certifi-2024.07.04, curl_cffi-0.7.1, mutagen-1.47.0, requests-2.32.3, sqlite3-3.45.3, urllib3-2.2.2, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets, curl_cffi
[debug] Loaded 1830 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-master-builds/releases/latest
Latest version: master@2024.08.19.221301 from yt-dlp/yt-dlp-master-builds
yt-dlp is up to date (master@2024.08.19.221301 from yt-dlp/yt-dlp-master-builds)
[ApplePodcasts] Extracting URL: https://podcasts.apple.com/tw/podcast/ep480/id1500839292?i=1000665698691
[ApplePodcasts] 1000665698691: Downloading webpage
ERROR: 1000665698691: An extractor error has occurred. (caused by StopIteration()); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp/extractor/common.py", line 740, in extract
  File "yt_dlp/extractor/applepodcasts.py", line 48, in _real_extract
  File "yt_dlp/utils/_utils.py", line 2640, in try_get
  File "yt_dlp/utils/_utils.py", line 2631, in try_call
  File "yt_dlp/extractor/applepodcasts.py", line 50, in <lambda>
StopIteration
```

**Comment by donaldguy** [comment]
Of significant note, Apple launched a usable web interface yesterday:

https://9to5mac.com/2024/08/19/apple-podcasts-web-app/

**Comment by donaldguy** [comment]
For a semi-randomly picked podcast episode, latest of the Daily (`https://podcasts.apple.com/us/podcast/biden-leaves-the-stage/id1200361736?i=1000665985607`), network traffic apparently took episode ID [1000665985607]:


and hit a REST API


`https://amp-api.podcasts.apple.com/v1/catalog/us/podcast-episodes/1000665985607?include=channel,podcast&include[podcasts]=episodes,podcast-seasons,trailers&include[podcast-seasons]=episodes&fields=artistName,artwork,assetUrl,contentRating,description,durationInMilliseconds,episodeNumber,guid,isExplicit,kind,mediaKind,name,offers,releaseDateTime,season,seasonNumber,storeUrl,summary,title,url&with=entitlements&l=en-US`

which:

- naively 401s.
- adding back bearer token [^1] not sufficient to avoid a 401. 
- also adding back `media-user-token` as header (vs original inclusion as head and cookie) was also not sufficient to escape. (base64 of this value does not decode to anything I or `file` could recognize)
- A full `Copy as curl`[^2] from the browser _did_ replay 
    - both with all the get params and 
    - cutting it down to `https://amp-api.podcasts.apple.com/v1/catalog/us/podcast-episodes/1000665985607?with=entitlements&l=en-US`

The latter returns somewhat manageable json blob

`{"data":[{"id":"1000665985607","type":"podcast-episodes","href":"/v1/catalog/us/podcast-episodes/1000665985607?l=en-US","attributes":{"offers":[{"kind":"get","type":"STDQ"}],"contentAdvisory":"","copyright":"© 2020-2021 THE NEW YORK TIMES COMPANY; The New York Times encourages the use of RSS feeds for personal use in a news reader or as part of a non-commercial blog, subject to your agreement to our Terms of Service.","genreNames":["Daily News"],"artworkOrigin":"episode","itunesTitle":"Biden Leaves the Stage","kind":"full","mediaKind":"audio","description":{"standard":"On the first night of the Democratic National Convention, the stage belonged to the man who chose to give it up.\n\nKatie Rogers and Peter Baker, White House correspondents for The Times, discuss President Biden’s private pain since stepping aside, and his public message in Chicago.\n\nGuest: \n\nKatie Rogers, a White House correspondent for The New York Times.\nPeter Baker, the chief White House correspondent for The New York Times.","short":"On the first night of the Democratic National Convention, the stage belonged to the man who chose to give it up.\n\nKatie Rogers and Peter Baker, White House correspondents for The Times, discuss President Biden’s private pain since stepping aside, and"},"artwork":{"width":3000,"height":3000,"url":"https://is1-ssl.mzstatic.com/image/thumb/Podcasts221/v4/9f/10/58/9f10584c-a79a-474e-c2c1-ad2fd17bef99/mza_10441190475887192964.jpg/{w}x{h}bb.{f}","bgColor":"100604","textColor1":"cac7c9","textColor2":"cbbfbb","textColor3":"a5a1a1","textColor4":"a59a96"},"url":"https://podcasts.apple.com/us/podcast/biden-leaves-the-stage/id1200361736?i=1000665985607","releaseDateTime":"2024-08-20T09:51:33Z","websiteUrl":"https://www.nytimes.com/the-daily","durationInMilliseconds":2012000,"name":"Biden Leaves the Stage","guid":"64e241f4-2ab0-4e37-8697-f67f3e4e5b2f","contentRating":"clean","artistName":"The New York Times","subscribable":true,"assetUrl":"https://dts.podtrac.com/redirect.mp3/chrt.fm/track/8DB4DB/pdst.fm/e/pfx.vpixl.com/6qj4J/nyt.simplecastaudio.com/03d8b493-87fc-4bd1-931f-8a8e9b945d8a/episodes/a4a1396a-8804-446a-9fbb-36170d81efec/audio/128/default.mp3?aid=rss_feed&awCollectionId=03d8b493-87fc-4bd1-931f-8a8e9b945d8a&awEpisodeId=a4a1396a-8804-446a-9fbb-36170d81efec&feed=54nAGcIl"},"relationships":{"artists":{"href":"/v1/catalog/us/podcast-episodes/1000665985607/artists?l=en-US","data":[{"id":"121664449","type":"artists","href":"/v1/catalog/us/artists/121664449?l=en-US"}]}}}]}`

of which the most interesting field is apparently `assetUrl` because the browser then (after a PUT at `/v1/me/playback/positions/podcast-episodes/100066598327`) did start making requests through that chain, but they were 302ing for CSP/CORS/ublock reasons

but ultimately there is a 206 (ranged okay) of mpeg data from `https://nyt.simplecastaudio.com/03d8b493-87fc-4bd1-931f-8a8e9b945d8a/episodes/a4a1396a-8804-446a-9fbb-36170d81efec/audio/128/default.mp3/default.mp3_ywr3ahjkcgo_414421b6245d9ab6cadbc39d53656d10_35528014.mp3?…`

this is interesting as it implies strongly that its not coming from Apple's CDN but the actual upstream of the podcast.


[^2]: in addition to Authorization and media-user-token, headers include:
    > Host: amp-api.podcasts.apple.com
    > User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:129.0) Gecko/20100101 Firefox/129.0
    > Accept: */*
    > Accept-Language: en-US,en;q=0.5
    > Accept-Encoding: gzip, deflate, br, zstd
    > Referer: https://podcasts.apple.com/
    > Origin: https://podcasts.apple.com
    > Connection: keep-alive
    > Sec-Fetch-Dest: empty
    > Sec-Fetch-Mode: cors
    > Sec-Fetch-Site: same-site
    > Priority: u=0
    > TE: trailers
    
    plus 23 cookies   
    some with apparently not sensitive key/value:
      
      1. dc=st
      1. itspod=7 (in fact repeated twice, for 24 actual cookies in the header)
      1. itua=US
      1. mut-refresh=1
      1. itre=0
      1. geo=US
      1. at_check=true
      1. s_cc=true
      1. dslang=US-EN
      1. site=usa

    keys with maybe/maybe-not sensitive hex values:
    
      1. `asrec` (much longer than)
      1. pltvcid
      1. pldfltcid 
    
    keys with urlencoded structured values:
    
      1. s_ppvl : 9 url-encoded comma separated values: 1st is double url encoded `acs::kb::bk::bk_[a hex string]_[a chip id?] (en-US)`, 2-7 are integers, 8th is a float, 9th is a letter; 
      2. `pt-dm` ; contained apparently 6 tilde seperated values, the last of which was "macbook air m1 - overview (tm)" [with the spaces as `%20`s]. I do have an m1 air, but was using windows at the time of this request.
      3. `s_sq` contained `[[B]]`
      4. `ac_history` contained the json `{"search":[],"kb":[["SP655","iPhone 5 - Technical Specifications","en_US",1693083602000,"unknown"],["SP655","iPhone 5 - Technical Specifications","en_US",1693081790000,"unknown"],["SP655","iPhone 5 - Technical Specifications","en_US",1693081701000,"unknown"]],"help":[],"psp":[],"offer_reason":{},"total_count":{"kbs":3,"last_kb":1693083602000}}`
      5. `mk_epub` also contains json with keys "btuid", "events" (contains a series of k=v pairs of numbered events with ... durations?) and "prop57":"www.tm.macbookairm1" (I thought that was the cancer warning one but no that's 65; CA prop 57 was about parole... so its probably not that either)
    
    then there are some with apparently-but-maybe-not-base64 encoded stuff that seems more likely sensitive:
    
      1. the aforementioned `media-user-token`
      2. `as_cn`
      3. `as_ltn_us`
      4. `as_pcts`
      5. `myacinfo` (which for middle 1291 chars is hex digits but has some letters at start and end; possibly notably two instances of `V3`) 


[^1]: containing a JWT with header alg `ES256` and `kid` (key id) `C4J7GBP74H`; payload, in addition to `exp` and `iat` has `iss` (Issuer) `UM7N9RTT7G` and field `root_https_origin` with array value containing `["apple.com"]`

looking at the network activity (for particularly a "Subscriber Edition" podcast I'd kinda hoped I could scrape out and load into PocketCasts), it is using a variation of the musickit stuff to pull chunk-at-a-time of an m3u8

**Comment by donaldguy** [comment]
As for @sharkshark123 's example from logs, the API returns assetUrl of `https://chtbl.com/track/4B4E1D/rss.soundon.fm/rssf/954689a5-3096-43a4-a80b-7810b219cef3/feedurl/c6d76d22-f777-4c01-af7a-d793fe4c0f7f/rssFileVip.mp3?timestamp=1723879937952`

which eventually becomes https://filesb.soundon.fm/file/filesb/2c3f9587-afa4-48b0-8c45-d0e87976d6b7.mp3

which indeed can be `wget`ed in its entirety directly

**Comment by coreywright** [comment]
this should be included in the next release after 2024.09.27.  until then you can use [nightly or master builds](https://github.com/yt-dlp/yt-dlp/blob/master/README.md#update) or [build from source](https://github.com/yt-dlp/yt-dlp/blob/master/README.md#compile).  (feel free to blame me and my inability to address PR comments until the following weekend, the day after 2024.09.27 was released).

**Comment by rainrdx** [comment]
> Meanwhile I had been hoping the launch of this web platform might help me exfiltrate a ~premium subscribed-to podcast from Apple and get it into a preferred pod catcher, but seems like no dice.
> 
> Once the assetUrl there gets unrolled it comes down to `https://play.itunes.apple.com/WebObjects/MZPlay.woa/podcast/hls/playlist.m3u8` with `cc=US`, `l=en-US`, `a` (the podcast's "album"?) and `id` (the albums track)
> 
> this breaks down into 7 more m3u8s on `https://podcasts.itunes.apple.com/itunes-assets/HLSPodcast211/v4/[first 2 hex digits of uuid]/[second two]/[third two]/[fourth two]/[a uuid]/P899701762_Anull_PC_en_audio_gr256_mp4a-40-2-.m3u8`
> 
> (which in turn point at segments that are `P899701762_Anull_PC_en_audio_gr256_mp4a-40-2-N.m4s` up to 118 on the first one I looked at)
> 
> the numbers after the P do seem to vary by podcast but not in a way clearly related to IDs used in URL or returned from that API (but does match the id param on that fetch)
> 
> passing in the top-level `play.` url to yt-dlp just goes:
> 
> ```
> [generic] playlist: Downloading webpage
> Extracting cookies from firefox
> Extracted 34 cookies from firefox
> [generic] playlist: Downloading m3u8 information
> [generic] playlist: Checking m3u8 live status
> [info] playlist: Downloading 1 format(s): 277
> [hlsnative] Downloading m3u8 manifest
> ERROR: This format is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
> ```
> 
> (doing `--list-formats` does in fact suggest there are some variations in bitrate offered here, but I imagine they are all equally DRMed)
> 
> and ditto for one of the `podcasts.` sub-playlists
> 
> ```
> [generic] P899701762_Anull_PC_en_audio_gr256_mp4a-40-2-: Downloading webpage
> [generic] P899701762_Anull_PC_en_audio_gr256_mp4a-40-2-: Downloading m3u8 information
> [generic] P899701762_Anull_PC_en_audio_gr256_mp4a-40-2-: Checking m3u8 live status
> ERROR: [generic] P899701762_Anull_PC_en_audio_gr256_mp4a-40-2-: This video is DRM protected
> ```
> 
> it seems like the decryption key must be being served somewhere in there (whether as comments/annotations in the m3u8s, of which there are many; or side-channeled)
> 
> while listening on the website, it continues to also POST intermittently to `https://amp-api.podcasts.apple.com/v1/me/playback/positions/podcast-episodes/[episode_id]?with=entitlements&l=en-US`
> 
> this might just be for ~sync (something the desktop and iOS clients are weirdly bad at), but they _might_ also be doing a check for faster-than-playback fetches?
> 
> (Though headers suggest that the `podcasts.itunes…/iunes-assets` are maybe going to Amazon CloudFront, and do they want to bother with CloudTrail ... stuff)

Thinking the same, but after tinkering around with this and gamdl seems no avail. Do hope to find ways to extract some premium podcasts..
